### PR TITLE
stop non-differentiable values from being materialized in aotautograd

### PR DESCRIPTION
**Pull request by Chillee**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #110773
* __->__ #110721
* #110720

### CODE REVIEW

**Comment by Chillee** [comment]
Interestingly this seems to have a significant impact in our benchmarks

image

**Comment by ezyang** [comment]
This seems plausible but you still have test errors

**Comment by Chillee** [comment]
Where do you see test errors 🤔

**Comment by ezyang** [comment]
I'm letting Brian review this

File: torch/_functorch/aot_autograd.py [hunk 1]
```diff
@@ -3175,56 +3181,46 @@ def backward(ctx, *flat_args):
...
+            # input_info contains info on *every* input,
+            # But in the backward(), we are only given grad outputs for every mutated input
+            # We then need to filter out the grad outputs that correspond to metadata-only mutations or don't require grad
+            mutated_inp_indices = CompiledFunction.metadata.mutated_inp_indices
+            input_info = CompiledFunction.metadata.input_info
+            assert len(inp_tangents) == len(mutated_inp_indices)
+            inp_tangents_filtered = [
+                x
+                for x, info_idx in zip(inp_tangents, mutated_inp_indices)
+                if input_info[info_idx].mutates_data and input_info[info_idx].requires_grad
```
Comment (bdhirsh) [inline torch/_functorch/aot_autograd.py:3199]: Hmm, I think the bit of the filter on `input_info[info_idx].requires_grad should be unnecessary now?

Previously, `inp_tangents` corresponded to "every forward input that had a (data or metadata) mutation". We had to filter this down to the actual inputs to our backward graph that correspond to user forward inputs, which are **fw inputs that had a data mutation and require grad**. (Actually - double checking, I think we're already correctly filtering out metadata mutations).

But now that you're filtering down to fw inputs that have a mutation and require grad even earlier (as part of constructing the joint fw/bw), we should only have to filter out inputs with metadata-only mutations in this check.

**Review by bdhirsh — APPROVED** [review]

File: torch/_functorch/aot_autograd.py (hunk 1 above)
Comment (Chillee) [reply torch/_functorch/aot_autograd.py:3199]: For future discussion, this isn't true. In this case, `inp_tangents` represents *all* of the inputs to the backwards pass, which will include values that corresponds to tensors that we don't actually trace with (for example, outputs of the forwards pass that are nondifferentiable).

So we still need to filter them out in this place - the logic is pretty analogous to the existing ones.

**Pull request closed without merging**